infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -172,42 +234,33 @@ private LogData readRecord(FileHandle fh, long address)
                 return null;
             }
 
-            int checksum = o.getInt();
-            ByteBuffer checksumBuf = o.slice();
-            long entryAddress = o.getLong();
-            int entryLen = o.getInt();
-
-            if (!noVerify) {
-                Hasher computedChecksum = Hashing.crc32c().newHasher();
+            int recordSize = o.getInt();
+            byte[] recordBuf = new byte[recordSize];
+            o.get(recordBuf);
 
-                int recordLen = Long.BYTES     // Size of address
-                        + Integer.BYTES  // Size of entry length
-                        + entryLen;      // entry size
-
-                for (int x = 0; x < recordLen; x++) {
-                    computedChecksum.putByte(checksumBuf.get());
-                }
+            try {
+                ChecksummedRecord checksummedRecord = ChecksummedRecord.parseFrom(recordBuf);
+                LogRecord record = checksummedRecord.getLogRecord();
 
-                if (checksum != computedChecksum.hash().asInt()) {
-                    log.error("Checksum mismatch detected while trying to read address {}", address);
-                    throw new DataCorruptionException();
+                if (!noVerify) {
+                    if (checksummedRecord.getChecksum() != getChecksum(record.toByteArray())) {
+                        log.error("Checksum mismatch detected while trying to read address {}", address);
+                        throw new DataCorruptionException();
+                    }
                 }
-            }
 
+                if (address == -1) {
+                    //Todo(Maithem) : maybe we can move this to getChannelForAddress
+                    fh.knownAddresses.add(record.getAddress());
+                } else if (record.getAddress() != address) {
+                    log.trace("Read address {}, not match {}, skipping. (remain={})", record.getAddress(), address);
+                } else {
+                    log.debug("Entry at {} hit, reading (size={}).", address, recordSize);
 
-            if (address == -1) {
-                //Todo(Maithem) : maybe we can move this to getChannelForAddress
-                fh.knownAddresses.add(entryAddress);
-            }
-
-            if (entryAddress != address) {
-                o.position(o.position() + entryLen); //skip over (size-20 is what we haven't read).
-                log.trace("Read address {}, not match {}, skipping. (remain={})", entryAddress, address, o.remaining());
-            } else {
-                log.debug("Entry at {} hit, reading (size={}).", address, entryLen);
-
-                ByteBuf buf = Unpooled.wrappedBuffer(o.slice());
-                return new LogData(buf);
+                    return getLogData(record);
+                }
+            } catch (InvalidProtocolBufferException e) {
```

Comment:
The file itself has a version right? So maybe we need a checksum for logheader, and that would be it?